File: 15-Graph/Topo Sort and Problems/01-Topological-Sort.cpp
/*
GFG:https://practice.geeksforgeeks.org/problems/topological-sort/1
Leetcode:
*/
#include <bits/stdc++.h>
using namespace std;
class bfsSolution // Kahn's Algorithm
{
public:
    // Function to return list containing vertices in Topological order.
    vector<int> topoSort(int V, vector<int> adj[])
    {
        vector<int> inDegree(V, 0);
        vector<int> ans;
        queue<int> q;

        for (int i = 0; i < V; i++)
            for (auto x : adj[i])
                inDegree[x]++;

        for (int i = 0; i < V; i++)
            if (!inDegree[i])
                q.push(i);

        while (!q.empty())
        {
            int node = q.front();
            ans.push_back(node);
            q.pop();

            for (auto x : adj[node])
            {
                inDegree[x]--;
                if (!inDegree[x])
                    q.push(x);
            }
        }
        return ans;
    }
};
class dfsSolution
{
private:
    void dfs(int i, vector<int> adj[], stack<int> &st, vector<int> &vis)
    {
        vis[i] = 1;
        for (auto x : adj[i])
            if (!vis[x])
                dfs(x, adj, st, vis);
        st.push(i);
    }

public:
    // Function to return list containing vertices in Topological order.
    vector<int> topoSort(int V, vector<int> adj[])
    {
        vector<int> vis(V, 0);
        vector<int> ans;
        stack<int> st;

        for (int i = 0; i < V; i++)
            if (!vis[i])
                dfs(i, adj, st, vis);

        while (!st.empty())
        {
            ans.push_back(st.top());
            st.pop();
        }
        return ans;
    }
};
int main()
{
}

File: 15-Graph/15.3-Topo Sort and Problems/01-Topological-Sort.py
class dfsSolution:
    # Function to return list containing vertices in Topological order.
    def dfs(self, i, adj, vis, stack):
        vis[i] = 1

        for x in adj[i]:
            if vis[x] == False:
                self.dfs(x, adj, vis, stack)
        stack.append(i)

    def topoSort(self, V, adj):
        # Code here
        vis = [0] * V

        stack = []

        for i in range(V):
            if vis[i] == False:
                self.dfs(i, adj, vis, stack)

        return stack[::-1]


from collections import deque


class kahnSolution:
    # Function to return list containing vertices in Topological order.
    def topoSort(self, V, adj):
        indegree = [0] * V
        queue = deque()

        for i in range(V):
            for node in adj[i]:
                indegree[node] += 1

        for i in range(V):
            if indegree[i] == 0:
                queue.append(i)
        ans = []
        while len(queue) != 0:
            node = queue.popleft()
            ans.append(node)

            for next in adj[node]:
                indegree[next] -= 1

                if indegree[next] == 0:
                    queue.append(next)

        return ans
